jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/DataUtils.scala

```diff
@@ -73,12 +74,81 @@ object DataUtils extends Serializable {
     }
   }
 
+  private def featureValueOfDenseVector(rowHashCode: Int, features: DenseVector): Float = {
+    val featureId = {
+      if (rowHashCode > 0) {
+        rowHashCode % features.size
+      } else {
+        // prevent overflow
+        math.abs(rowHashCode + 1) % features.size
+      }
+    }
+    features.values(featureId).toFloat
+  }
+
+  private def featureValueOfSparseVector(rowHashCode: Int, features: SparseVector): Float = {
+    val featureId = {
+      if (rowHashCode > 0) {
+        rowHashCode % features.indices.length
+      } else {
+        // prevent overflow
+        math.abs(rowHashCode + 1) % features.indices.length
+      }
+    }
+    features.values(featureId).toFloat
+  }
+
+  private def calculatePartitionKey(row: Row, numPartitions: Int): Int = {
+    val Row(_, features: Vector, _, _) = row
+    val rowHashCode = row.hashCode()
+    val featureValue = features match {
+      case denseVector: DenseVector =>
+        featureValueOfDenseVector(rowHashCode, denseVector)
+      case sparseVector: SparseVector =>
+        featureValueOfSparseVector(rowHashCode, sparseVector)
+    }
+    (rowHashCode.toLong + featureValue).toInt % numPartitions
```

Comment:
> I still do not understand why do we need to add featureValue to already "quite random" hash code?

you cannot guarantee row.hashCode is conflict-free, if you go into the details of how MurmurHash3 is implemented. The conflict will come from two parts: (1) hashcode conflict of the column value; (2) the lower 26 bits conflicts in the mixed value of 'Seq' and the hash code of column value;

I didn't see any problem by bring an additional 32 bits to avoid conflict

>  I am a bit concern that featureValue + rowHashCode could be more then MAX_INT.

I made a mistake here, I will address it